New tutorial using interactive API on the brainMRI segmentation dataset

saransh09 opened the pull request:
* A new segmentation tutorial using the latest version of Interactive API of OpenFL
* Using the Medical Decathlon BrainMRI dataset
* Tried to make the tutorial more informative and dumbed it down a little to make it smoother to use for general devolopers
* Till now most of the tutorials were using a dataset which was very small (in the order of 50-60MBs)
* This dataset used >10GB, will be good to demonstrate how OpenFL scales and can be used on larger datasets
* has a custom shard_descriptor which is quite different from the KVASIR example (as the data is organised very differently)
* Using CSVs to keep a track of the data

saransh09 committed:
[PyTorch_BrainMRI] new tutorial added

saransh09 committed:
[README.md] added README.md file and updated simething in prepare_dataset.py

openfl-helper commented:
Can one of the admins verify this patch?

alexey-khorkin commented on openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/prepare_dataset.py:
```diff
@@ -0,0 +1,128 @@
+os.system('cp processed_data/train_data.csv envoy_1/')
+os.system('cp processed_data/train_data.csv envoy_2/')
+os.system('cp processed_data/valid_data.csv envoy_1/')
+os.system('cp processed_data/valid_data.csv envoy_2/')
```
@saransh09 , can we refactor code to add the following at the end of the file?
```python3
if __name__ == '__main__':
    main()
```

saransh09 replied:
> @saransh09 , can we refactor code to add the following at the end of the file?
> 
> ```python
> if __name__ == '__main__':
>     main()
> ```

Yes sure, I will do this, on it.

alexey-khorkin commented:
@saransh09 do we really need `envoy_2` folder? Please, check other examples.

alexey-khorkin commented on openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy_1/shard_descriptor.py:
```diff
@@ -0,0 +1,193 @@
+# Copyright (C) 2020-2021 Intel Corporation
```
@saransh09 could we add type annotations? Like here - https://github.com/intel/openfl/blob/develop/openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py

saransh09 replied:
Sure, I am on it.

alexey-khorkin reviewed (changes requested):
Need changes

saransh09 commented:
@alexey-khorkin 
I deliberately kept the envoy_2 folder because I wanted the two envoy setup to be clearer. However, I understand that to maintain the homogeneity of the examples. I will remove it. There is one problem here, that the data is present in google drive and there is no way to download it through the link. The two folders (for envoys) were made so that the script would process and copy the data in the respective folders as needed. Else, it would become a little too opaque to run the experiment for multiple envoys

alexey-gruzdev commented:
@saransh09 could you please drop a comment with signing CLA - `I have read the CLA Document and I hereby sign the CLA`, otherwise next stages with Jenkins build are skipped.

saransh09 commented:
I have read the CLA Document and I hereby sign the CLA

alexey-gruzdev commented:
ok to test

saransh09 committed:
[adding type annotations]

alexey-gruzdev commented:
Looks like CI is not green, I believe you need first fix - [Linter](http://213.221.44.203/job/Federated-Learning/job/precommit/1496/execution/node/32/log/)

saransh09 commented:
> Looks like CI is not green, I believe you need first fix - [Linter](http://213.221.44.203/job/Federated-Learning/job/precommit/1496/execution/node/32/log/)

Thanks for this, reading the comments and changing everything. I will make all the changes and make another PR in some time

alexey-gruzdev commented:
> > Looks like CI is not green, I believe you need first fix - [Linter](http://213.221.44.203/job/Federated-Learning/job/precommit/1496/execution/node/32/log/)
> 
> Thanks for this, reading the comments and changing everything. I will make all the changes and make another PR in some time

You don't need to open another PR. Just add fixes on top of your current branch.

alexey-khorkin commented:
> There is one problem here, that the data is present in google drive and there is no way to download it through the link. 


@saransh09  we use `gdown` to download files from google drive, for example: https://github.com/intel/openfl/blob/7af3fc8f0ac0328e19fb7b407c36525edb1506ae/openfl-tutorials/interactive_api/Tensorflow_Word_Prediction/envoy/shard_descriptor.py#L137

saransh09 commented:
> > There is one problem here, that the data is present in google drive and there is no way to download it through the link.
> 
> @saransh09 we use `gdown` to download files from google drive, for example:
> 
> https://github.com/intel/openfl/blob/7af3fc8f0ac0328e19fb7b407c36525edb1506ae/openfl-tutorials/interactive_api/Tensorflow_Word_Prediction/envoy/shard_descriptor.py#L137

Done. Added it and refactored it into the shard_descriptor to be consistent to the other tutorials. Thanks for the suggestion.

saransh09 committed:
[refactoring the code in accordance to the other tutorials]

saransh09 committed:
Merge branch 'intel:develop' into saransh_new_tutorial

saransh09 committed:
Merge branch 'intel:develop' into saransh_new_tutorial

saransh09 committed:
[linting flake8]

saransh09 committed:
Merge branch 'saransh_new_tutorial' of https://github.com/saransh09/openfl-1 into saransh_new_tutorial
Adding the updated changes from develop branch

saransh09 committed:
[adding Copyright to missing files]

alexey-khorkin reviewed (changes requested):
Please, remove data files from PR: 
* `openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/train_data.csv`
* `openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/valid_data.csv`

saransh09 committed:
[removing csv files]

saransh09 commented:
> Please, remove data files from PR:
> 
> * `openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/train_data.csv`
> * `openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/valid_data.csv`

Done!

saransh09 commented:
The Jenkins pipeline has been running for ~24 hours now. Is there something wrong from my end?

itrushkin commented:
> The Jenkins pipeline has been running for ~24 hours now. Is there something wrong from my end?

Looks like CI is stuck. There is an issue on Jenkins side. Restarting the build.

Jenkins please retry a build

itrushkin commented:
Jenkins please retry a build

igor-davidyuk commented on openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/.gitignore:
```diff
@@ -0,0 +1,12 @@
+Task01_BrainTumour.tar
```
This file should be removed

igor-davidyuk commented on openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py:
```diff
@@ -0,0 +1,300 @@
+
+        return img[tuple(slices)], msk[tuple(slices)]
+
+    def augment_data(self, img, msk) -> Tuple[Any, Any]:
```
Augmenting data is a part of FL experiment. Shard descriptor just provides local data

igor-davidyuk commented on openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py:
```diff
@@ -0,0 +1,300 @@
+
+    def __init__(self, csv_path, rank=1, worldsize=1,
+                 crop_dims=None, dataset_dir='./',
+                 augment=True, device='cpu', set='train') -> None:
```
```suggestion
                 augment=True, set='train') -> None:
```
Managing devices also is no concern for a Shard Descriptor

igor-davidyuk commented on openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py:
```diff
@@ -0,0 +1,300 @@
+                self.csv = self.csv.loc[(self.rank - 1) * (len(
+                    self.csv) // self.worldsize):(self.rank) * (len(
+                        self.csv) // self.worldsize), :].reset_index(drop=True)
+        self.device = device
+        self.augment = augment
+
```
```suggestion


```

igor-davidyuk commented on openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py:
```diff
@@ -0,0 +1,300 @@
+        img = self.preprocess_img(img)
+        msk = self.preprocess_label(msk)
+        img, msk = self.crop_input(img, msk)
+        if self.augment:
+            img, msk = self.augment_data(img, msk)
+        img = img[np.newaxis, :, :].copy()
```
Augmenting should be moved to experiment's Dataloader part

mansishr commented on openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py:
```diff
@@ -0,0 +1,300 @@
+            print("processed_data/ already exists, do you want to proceed? \
+                press y/Y to proceed and n/N to cancel")
+            inp = input()
+            if inp == 'y' or inp == 'Y':
```
Processing images and masks shouldn't be under download_data function.

mansishr commented on openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/workspace/layers.py:
```diff
@@ -0,0 +1,102 @@
+# Copyright (C) 2020-2021 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+"""Layers for Unet model."""
+
```
demo images under workspace should be removed

mansishr commented on openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/workspace/BrainMRI2D_Segmentation_OpenFL_Tutorial.ipynb:
```diff
@@ -0,0 +1,476 @@
+    "import torch.nn.functional as F \n",
+    "import tqdm\n",
+    "from openfl.component.aggregation_functions import Median\n",
+    "from layers import soft_dice_coef, soft_dice_loss, DoubleConv, Down, Up"
```
soft_dice_coef and soft_dice_loss are being redefined in the notebook

mansishr commented on openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/workspace/BrainMRI2D_Segmentation_OpenFL_Tutorial.ipynb:
```diff
@@ -0,0 +1,476 @@
+    "### Download the dataset : Medicat Decathlon BrainMRI Segmentation Dataset\n",
+    "\n",
+    "Download the data from : https://goo.gl/QzVZcm download the file Task01_BrainTumour.tar (into the base repository) <p>\n",
+    "Then run the prepare_dataset.py file to process and create the dataset so that it is ready to use. <p>\n",
```
The instructions to download the dataset/setup etc should be a part of README and removed from the notebook.

theakshaypant commented:
@saransh09 can you please resolve the comments and request for review again?

psfoley commented:
Closing due to deprecation of the Interactive API
